contracts/protocol/libraries/logic/PositionMoverLogic.sol

```diff
@@ -41,103 +40,176 @@ library PositionMoverLogic {
 
-        lendPool.repay(loanData.nftAsset, loanData.nftTokenId, borrowAmount);
+        lendPool.batchRepay(
+            tmpVar.nftAssets,
+            tmpVar.tokenIds,
+            tmpVar.borrowAmounts
+        );
+
+        supplyAssets(ps, tmpVar);
 
-        (nftAsset, tokenId) = (loanData.nftAsset, loanData.nftTokenId);
+        borrowAmount = tmpVar.totalBorrowAmount;
     }
 
-    function supplyNFTandBorrowWETH(
+    function supplyAssets(
         DataTypes.PoolStorage storage ps,
-        IPoolAddressesProvider poolAddressProvider,
         PositionMoverVars memory tmpVar
     ) internal {
+        uint256 tokenIdsLength = tmpVar.tokenIds.length;
         DataTypes.ERC721SupplyParams[]
-            memory tokenData = new DataTypes.ERC721SupplyParams[](1);
-        tokenData[0] = DataTypes.ERC721SupplyParams({
-            tokenId: tmpVar.tokenId,
+            memory tokensToSupply = new DataTypes.ERC721SupplyParams[](
+                tokenIdsLength
+            );
+
+        address currentSupplyAsset = tmpVar.nftAssets[0];
+        uint256 supplySize = 1;
+        tokensToSupply[0] = DataTypes.ERC721SupplyParams({
+            tokenId: tmpVar.tokenIds[0],
             useAsCollateral: true
         });
+        uint256 nextIndex;
+
+        for (uint256 index = 0; index < tokenIdsLength; index++) {
+            nextIndex = index + 1;
+            if (
+                nextIndex < tokenIdsLength &&
+                tmpVar.nftAssets[index] == tmpVar.nftAssets[nextIndex]
+            ) {
+                tokensToSupply[supplySize] = DataTypes.ERC721SupplyParams({
+                    tokenId: tmpVar.tokenIds[nextIndex],
+                    useAsCollateral: true
+                });
+                supplySize++;
+            } else {
+                reduceArrayAndSupply(
+                    ps,
+                    currentSupplyAsset,
+                    tokensToSupply,
+                    supplySize
+                );
+
+                if (nextIndex < tokenIdsLength) {
+                    currentSupplyAsset = tmpVar.nftAssets[nextIndex];
+                    tokensToSupply = tokensToSupply = new DataTypes.ERC721SupplyParams[](
+                        tokenIdsLength
+                    );
+                    tokensToSupply[0] = DataTypes.ERC721SupplyParams({
+                        tokenId: tmpVar.tokenIds[nextIndex],
+                        useAsCollateral: true
+                    });
+                    supplySize = 1;
+                }
+            }
+        }
+    }
 
+    function reduceArrayAndSupply(
+        DataTypes.PoolStorage storage ps,
+        address asset,
+        DataTypes.ERC721SupplyParams[] memory tokensToSupply,
+        uint256 subArraySize
+    ) internal {
+        uint256 itemsToTrim = tokensToSupply.length - subArraySize;
+
+        if (itemsToTrim != 0 && subArraySize != 0) {
+            assembly {
+                mstore(tokensToSupply, sub(mload(tokensToSupply), itemsToTrim))
```

Comment:
could we use mstore(tokensToSupply, subArraySize)   and remove the field itemsToTrim